packages/esm-patient-test-results-app/src/filter/filter-set.tsx

```diff
@@ -22,43 +23,74 @@ const FilterSet: React.FC<FilterSetProps> = ({ hideFilterSetHeader = false }) =>
   const { t } = useTranslation();
   const { resetTree } = useContext(FilterContext);
   const [searchTerm, setSearchTerm] = useState('');
+  const [treeDataFiltered, setTreeDataFiltered] = useState(roots);
   const [showSearchInput, setShowSearchInput] = useState(false);
 
+  useEffect(() => {
+    setTreeDataFiltered(roots);
+  }, [roots]);
+
+  function filterTreeNode(inputValue, treeNode) {
+    // If the tree node's display value contains the user input, or any of its children's display contains the user input, return true
+    if (
+      treeNode?.display.toLowerCase().includes(inputValue.toLowerCase()) ||
+      (treeNode?.subSets && treeNode.subSets.some((child) => filterTreeNode(inputValue, child)))
+    ) {
+      return true;
+    }
+
+    // Otherwise, return false
+    return false;
+  }
+
+  function handleInputChange(value) {
```

Comment:
Either lift these functions out of the React component or wrap them in `useCallback()` please. (Functions defined as part of the body of a React functional component are re-evaluated on every render cycle; that means, each time this component re-renders, we're replacing, e.g., `filterTreeNode()` with a new `filterTreeNode()` function; this is basically the exact reason `useCallback()` exists).